File: python/077_saddle-points/saddle_points.py
def saddle_points(matrix):
  try:
    output = []
    row_count = len(matrix)
    col_count = len(matrix[0]) if matrix else 0
    max_row = [max(row) for row in matrix]
    min_col = [min([matrix[row][col] for row in range(row_count)]) for col in range(col_count)]
    for row in range(row_count):
        for col in range(col_count):
            if matrix[row][col] == max_row[row] and matrix[row][col] == min_col[col]:
                output.append({'row':row+1, 'column':col+1})
    return output
  except:
    raise ValueError("irregular matrix")

File: javascript/081_saddle-points/saddle-points.js
export const saddlePoints = (matrix) => {
  const maxs_in_rows = [];
  const mins_in_cols = [];
  const output = [];
  const max_in_row = row => {
    if(!maxs_in_rows[row])
      maxs_in_rows[row] = Math.max(...matrix[row]);
    return maxs_in_rows[row];
  };
  const min_in_col = col => {
    if(!mins_in_cols[col])
      mins_in_cols[col] = Math.min(...matrix.map(row => row[col]));
    return mins_in_cols[col];
  };
  matrix.forEach((row, i) => {
    row.forEach((tree, j) => {
      if(tree === max_in_row(i) && tree === min_in_col(j))
        output.push({row: i + 1, column: j + 1});
    });
  });
  return output;
};
